File: src/evn_postprocess/scripts/feedback.pl
<<<<<<< HEAD
#!/local/bin/perl -w
=======
#!/usr/bin/env perl
>>>>>>> pipe2
# Cormac Reynolds: February 2001 - original program
# Simple program to produce the NME feedback pages and provide links to the evn
# pipeline output postscript files
# 18 Dec 2003 - made separate links for each source to selfcal solutions.
# 21 Feb 2006 - link to .pdf instead of .PS.gz. No longer need to do any
#     conversion .
# 10 Dec 2014 - EPB, remove requirement on 'use Shell'
# 30 Sep 2022 - Disabling the mandatory ^D and RTN inputs from user

#use Shell ("date", "convert", "gzip", "gunzip");
use Getopt::Long;

print "use -exp option to input experiment name\n";
print "use -source option to input source names (use quotes for more than one
source). By default will use all sources with plot files in the pwd.\n";
#print "use -gzip=F to not gzip files and have links point to uncompressed PS ",
#      "files\n";
print "use -nme=T to format the page for NMEs, otherwise formatted for user experiments\n";
print "use -jss to specify the contact email to appear on the web page (omit
the '\@jive.eu')\n";
#print "use -pngok if you don't want to remake the png files\n";
print "experiments pipelined before 18 Dec 2003 should use feedback.pl.031218\n";

GetOptions('exp=s'=>\$expname, 'source=s'=>\$sources, 'nme=s'=>\$nme, 
          'jss=s'=>\$jss);
if ($sources) {
  $sources =~ s/^\s+//;
  @sources = split (/\s+/, $sources);
}

unless (defined $jss) {
  print "\n\nEnter your email address (omit the '\@jive.eu')\n";
  chomp ($jss = <STDIN>);
}
unless ($jss =~ /@/) {
  $jss = $jss . '@jive.eu';
}

$nme = uc $nme      if (defined $nme);

unless (defined $expname) {
  print "\n\nWhat is the experiment name?\n";
  chomp ($expname = <STDIN>);
}
unless (defined $sources) {
  @sourcesguess = getsources($expname);
  print "\n\nWhat sources are to be plotted?\n";
  print "By default, will use the following sources: @sourcesguess\n";
  # print "Enter just ^D if the default list is o.k.\n";
  # chomp (@sources = <STDIN>);
}

unless (@sources) {
  @sources = @sourcesguess;
}
foreach (@sources) {
  tr/a-z/A-Z/ ;
}
#nmedir is the directory that everything is in (psfiles, comment file, output
#html file etc.). wwwdir is the path name used
#for the links - it may be a relative path.
#print "\n\nEnter the directory containing the plot files (<RTN> for pwd):\n";
#chomp ($nmedir = <STDIN>);
#if ($nmedir eq '') {
$nmedir = '.';
#}
unless ($nmedir =~ /\/$/) {
  $nmedir = $nmedir . '/';
}
#$wwwdir = ("/juw08_1/scratch/cormac/nme/$expname/");
$wwwdir = ("./");
$htmlfile = ("$nmedir$expname.html");
#print "$htmlfile\n";
open (OUTFILE,">$htmlfile") || die "can\'t open file $!\n";
#opendir (NMEDIR, "$nmedir") || die "can\'t read the NME directory: $!\n";
#@psfiles = grep (/$expname.*\.PS$/, readdir NMEDIR);
#print "@psfiles\n";

#make the basic web page:
print OUTFILE <<EOT
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
        "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
EOT
;
if (defined $nme && $nme eq 'T') {
  print OUTFILE "<title>European VLBI Network NME Feedback</title>\n";
}else {
  print OUTFILE "<title>European VLBI Network Pipeline Feedback</title>\n";
}
print OUTFILE <<EOT
        <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
        <link rel="stylesheet"
              type="text/css"
              href="http://old.evlbi.org/css/evn.css"
              title="JIVE style sheet" />

EOT
;
print OUTFILE "</head>\n";
print OUTFILE "<body>\n";
print OUTFILE "\n";
print OUTFILE "\n";
print OUTFILE <<EOT
<div class="totale">


   <div class="content">

      <div class="banner">
                <script type="text/javascript"
                        src="http://old.evlbi.org/javascript/imagecycle4.js">
                </script>
                <noscript>
                <img src="http://old.evlbi.org/graphics/banner/evnbanner.gif" alt="EVN Banner" />
                </noscript>
      </div>

      <div class="main">
EOT
;
#print OUTFILE "<ul>\n";
print OUTFILE "\n";
print OUTFILE "\n";
print OUTFILE "<br />\n";
#print OUTFILE "<center>\n";
if (defined $nme && $nme eq 'T') {
  print OUTFILE "<h1 class=\"title\">EVN Network Monitoring Experiment (NME) Pipeline Feedback </h1>\n";
}else {
  print OUTFILE "<h1 class=\"title\">EVN User Experiment Pipeline Feedback </h1>\n";
}
print OUTFILE "<br />\n";
#print OUTFILE "</center>\n";
print OUTFILE "\n\n";
print OUTFILE "<p>\n";
print OUTFILE "<object>\n";
print OUTFILE "<center>\n";
print OUTFILE "<em>\n";
$expnameup = $expname;
$expnameup =~ tr/a-z/A-Z/;
print OUTFILE "Pipeline feedback for experiment $expnameup. \n";
print OUTFILE "If you have any comments on this experiment please email \n the
                  address below. <br />\n",
                  "A <a
               href=\"http://old.evlbi.org/pipeline/pipe_desc.html\">detailed
                  description of the pipeline output</a> is available. <br />
                  \n ";

print OUTFILE "</em>\n";
#print OUTFILE "The AIPS calibration tables produced by the pipeline can be
#               retrieved from the <a
#               href=\"ftp://vlbeer.ira.bo.cnr.it/vlbi_arch/feb02\">VLBEER ftp
#               server</a>\n";
print OUTFILE "</center>\n";
print OUTFILE "</object>\n";

print OUTFILE "</p>\n\n";
print OUTFILE "<hr />\n";
print OUTFILE "<p>\n";
print OUTFILE "<!-- HTML file last updated: -->\n";
$update = `date`;
print OUTFILE "Last updated: $update\n";
print OUTFILE "<a href=\"mailto:$jss\">$jss</a>\n";
print OUTFILE "\n";
#print OUTFILE "</table>\n";
print OUTFILE "<br /> \n";

#read in the comments file
$commfile = ("$nmedir$expname.comment");
#print "$commfile\n";
if (-e $commfile) {
  open (COMMFILE,"$commfile") ;
  $i=0;
  while (<COMMFILE>) {
    chomp;
    $comment[$i] = $_;
    ++$i;
  }
} else {
  $comment[0] = ' ';
  #print "comment=@comment\n"
}

$commjoin = join (' ', @comment);
@comment = split (/\/\/\//, $commjoin);
#print "@comment";

print OUTFILE "</p>\n\n";
print OUTFILE "<hr /> \n";



$name = "GENERAL";
print OUTFILE "<p>\n";
print OUTFILE "<font color=\"\#4400CC\">\n";
print OUTFILE "General Comments. \n";
print OUTFILE "</font>\n";
$link = ($wwwdir . $expname .  '.DTSUM');
($start_href, $end_href) = mkhref($link);
print OUTFILE "(";
print OUTFILE $start_href;
print OUTFILE "Brief data summary";
print OUTFILE $end_href;
print OUTFILE " and ";
$link = ($wwwdir . $expname .  '.SCAN');
($start_href, $end_href) = mkhref($link);
print OUTFILE $start_href;
print OUTFILE "scan listing";
print OUTFILE $end_href, ") <br />\n";
print OUTFILE getcomm($name);
#print OUTFILE "$thiscomment\n";
#print OUTFILE "<br /> \n";
print OUTFILE "<br /> \n</p>  \n";
#$i=0;
#while ($i<@comment) {
#  print "comment2=$comment[$i]\n";
#  ++$i;
#}

#print OUTFILE "<p>\n";
$name = "ERIVAL";
#print OUTFILE " The EVN reliability indicator (ERI) for this experiment was
#ERI = ", getcomm($name), ". ";
$name = "ERISTAR";
#print OUTFILE "ERI* = ", getcomm($name), ".\n";
$name = "ERICOMM";
#print OUTFILE getcomm($name), "\n";

#print OUTFILE "<br /> \n</p> \n<hr /> \n \n<p> \n";
print OUTFILE "\n<hr /> \n \n<p> \n";
#and put the links to the ps files

wrt_section("POSSM_AUTOCORR", "Plots of the autocorrelations\n");

wrt_section("VPLOT_UNCAL", "Plots of the uncalibrated amplitude and phase
against time");

wrt_section("POSSM_UNCAL", "Plots of the uncalibrated amplitude and phase
against frequency channel\n");

wrt_section("POSSM_CPOL", "The uncalibrated amplitude and phase of the crosshand
correlations against frequency channel\n");

wrt_section("TSYS", "TSYS against time\n");

wrt_section("GAIN", "Telescope sensitivities ", "from the a priori TSYS and Gain
curves (the square of this number gives the antenna noise (SEFD) in Jy - the
smaller the better).");

wrt_section("FRING_PHAS", "Fringe-fit phase solutions ", "(including Parallactic
Angle correction).");

wrt_section("FRING_DELAY", "Fringe-fit delay solutions\n");

wrt_section("FRING_RATE", "Fringe-fit rate solutions\n");

wrt_section("BANDPASS", "Telescope bandpasses\n");

wrt_section("VPLOT_CAL", "Calibrated amplitude and phase against time ", "(a
priori amplitude calibration and fringe-fit solutions applied).");

wrt_section("POSSM_CAL", "Calibrated amplitude and phase against frequency
channel\n");

$nmaps=0;
$name = "IMAPN";
print OUTFILE "Naturally weighted dirty map (not useful for bright sources)
produced before self-cal of: <br />";
while ($nmaps < @sources) {
  @suffix = ('pdf', 'FITS');
  print OUTFILE "$sources[$nmaps]:";
  wrt_map_section (\$name, \$nmaps, \@suffix, \@suffix) ;
  ++$nmaps;
}
if ($nmaps > 0) {
  wrcomm($name);
}

$nmaps=0;
$name = "IMAPU";
print OUTFILE "Uniformly weighted dirty map (not useful for bright sources)
produced before self-cal of: <br />";
while ($nmaps < @sources) {
  @suffix = ('pdf', 'FITS');
  print OUTFILE "$sources[$nmaps]:";
  wrt_map_section (\$name, \$nmaps, \@suffix, \@suffix) ;
  ++$nmaps;
}
if ($nmaps > 0) {
  wrcomm($name);
}

$name = "CALIB_PHAS1";
print OUTFILE "Phase corrections applied to a priori calibrated and
fringe-fitted data by self-calibration. <br />\n";
$nmaps = 0;
while ($nmaps < @sources) {
  @suffix = ('pdf');
  @text = ("$sources[$nmaps]");
  wrt_map_section (\$name, \$nmaps, \@suffix, \@text) ;
  ++$nmaps;
}
wrcomm($name);

$name = "CALIB_AMP2";
print OUTFILE "Amplitude corrections applied to a priori calibrated and
fringe-fitted data by self-calibration. <br />\n";
$nmaps = 0;
while ($nmaps < @sources) {
  # do the amplitude calibration summary
  $link = ($wwwdir . $expname . "_" . $sources[$nmaps] . '_' . $name . ".TXT" );
<<<<<<< HEAD
  if (-d "/aps3/Pypeline/reynolds"){
    #system "~reynolds/progs/perl/ampcalstats.pl $link" ;
    system "/aps3/Pypeline/reynolds/perl/ampcalstats.pl $link" ;
  }
=======
  #if (-d "/aps3/Pypeline/reynolds"){
    #system "~reynolds/progs/perl/ampcalstats.pl $link" ;
  system "/home/jops/opt/evn_support/ampcalstats.pl $link" ;
  #}
>>>>>>> pipe2
  @suffix = ('pdf', 'TXT', 'ampcal');
  @text = ('pdf', 'text file', 'statistical summary');
  print OUTFILE "$sources[$nmaps]:";
  wrt_map_section (\$name, \$nmaps, \@suffix, \@text) ;
  ++$nmaps;
}
wrcomm($name);

$name = "SENS";
wrt_section("SENS", "Telescope sensitivities", " (the total AMP gain applied
during both a priori and self calibration; the square of this number gives the
antenna noise (SEFD) in Jy).\n");

$nmaps=0;
$name = "CLPHS";
print OUTFILE "Residual closure phase (visibility closure phase with model",
            " closure phase subtracted) for: <br />";
while ($nmaps < @sources) {
  @suffix = ('pdf');
  @text = ("$sources[$nmaps]");
  #print OUTFILE "$sources[$nmaps]:";
  wrt_map_section (\$name, \$nmaps, \@suffix, \@text) ;
  ++$nmaps;
}
if ($nmaps > 0) {
  wrcomm($name);
}

$nmaps=0;
$name = "VPLOT_MODEL";
print OUTFILE "Calibrated visibilities and the source model of: <br />";
while ($nmaps < @sources) {
  @suffix = ('pdf');
  @text = ("$sources[$nmaps]");
  wrt_map_section (\$name, \$nmaps, \@suffix, \@text) ;
  ++$nmaps;
}
if ($nmaps > 0) {
  wrcomm($name);
}

$nmaps=0;
$name = "UVPLT";
print OUTFILE "Calibrated visibilities against <em>u,v</em> distance for: <br />";
while ($nmaps < @sources) {
  @suffix = ('pdf', 'png');
  print OUTFILE "$sources[$nmaps]:";
  wrt_map_section (\$name, \$nmaps, \@suffix, \@suffix) ;
  ++$nmaps;
}
if ($nmaps > 0) {
  wrcomm($name);
}

$nmaps=0;
$name = "UVCOV";
print OUTFILE "<em>u,v</em> coverage for: <br />";
while ($nmaps < @sources) {
  @suffix = ('pdf', 'png');
  print OUTFILE "$sources[$nmaps]:";
  wrt_map_section (\$name, \$nmaps, \@suffix, \@suffix) ;
  ++$nmaps;
}
if ($nmaps > 0) {
  wrcomm($name);
}

$nmaps=0;
$name = "ICLN";
print OUTFILE "<em>Crude</em> maps of sources: <br />";
while ($nmaps < @sources) {
  @suffix = ('pdf', 'FITS');
  print OUTFILE "$sources[$nmaps]:";
  wrt_map_section (\$name, \$nmaps, \@suffix, \@suffix) ;
  ++$nmaps;
}
if ($nmaps > 0) {
  wrcomm($name);
}


print OUTFILE "</p>\n\n";
print OUTFILE "\n";
print OUTFILE "\n";
#print OUTFILE "</center>\n";
#print OUTFILE "</ul>\n";
print OUTFILE "</div>\n";
print OUTFILE <<EOT
<!--#include virtual="./foot.html"-->

EOT
;
print OUTFILE "</div>\n";
print OUTFILE "</div>\n";

print OUTFILE "</body>\n";
print OUTFILE "</html>\n";

chmod 0755, $htmlfile;


#sub mkimg {
#  my ($psname) = $_[0];
#  my ($pngname) = $psname;
#  $pngname =~ s/PS/png/;
#  if (-e $psname) {
#    print "converting $psname to $pngname\n";
#    #pstoimg ("$psname");
#    convert ("$psname $pngname");
#  }elsif (-e "$psname.gz") {
#    gunzip ("$psname.gz") ;
#    print "converting $psname.gz to $pngname\n";
#    #pstoimg ("$psname");
#    convert ("$psname $pngname");
#  }
#}

#sub zipit {
#  my $file = $_[0];
#  if ($gz) {
#    if (-e $file) {
#      print "gzipping $file\n";
#      gzip ("$file")  ;
#    }
#    $file .= $gz;
#  }
#  return $file;
#}

sub mkhref {
  my ($link) = $_[0];
  my ($start_href, $end_href) ;
  if (-e $link) {
    $start_href = "<a href =\"$link\">\n";
    $end_href = "</a> \n";
  }else {
    $start_href = "";
    $end_href = "<font color=\"\#FF0000\"> (not available) </font> \n";
  }
  return $start_href, $end_href;
}

sub wrcomm {
  my ($name) = $_[0];
  print OUTFILE "<font color=\"\#4400CC\">\n";
  print OUTFILE "Comments. <br /> \n";
  print OUTFILE "</font>\n";
  $thiscomment = getcomm($name);
  #@thiscomm = grep (/^ *$name/, @comment);
  #$thiscomment = join (' ', @thiscomm);
  #$thiscomment =~ s/^ *$name//g;
  print OUTFILE "$thiscomment\n";
  print OUTFILE "<br /> \n</p> \n<hr /> \n \n<p> \n";
}

sub getsources {
  my ($expname) = $_[0];
  opendir (THISDIR, ".") || die "can\'t read the current directory: $!\n";
  my (@sourcesguess) = grep (/$expname.*_UVPLT.pdf/, readdir THISDIR);
  foreach my $source (@sourcesguess) {
    $source =~ s/$expname\_(.*)_UVPLT.pdf.*/$1/;
  }
  closedir (THISDIR);
  return @sourcesguess;
}

sub getcomm {
  my ($name) = $_[0];
  @thiscomm = grep (/^ *$name/, @comment);
  $thiscomment = join (' ', @thiscomm);
  $thiscomment =~ s/^ *$name//g;
  return $thiscomment;
}

sub wrt_section{
  my ($name, $text, $text2) = @_;
  $link = ($wwwdir . $expname . "_" . $name . ".pdf" );
  #$link=zipit($link);
  ($start_href, $end_href) = mkhref($link);
  print OUTFILE $start_href;
  print OUTFILE $text;
  print OUTFILE $end_href;
  print OUTFILE $text2 if $text2;
  print OUTFILE "<br />\n";
  wrcomm($name);
}


sub wrt_map_section{
  my ($name, $nmaps, $suffix, $text) = @_;
  $name = ${$name};
  $nmaps = ${$nmaps};
  @suffix = @{$suffix};
  @text = @{$text};
  for (my $i=0; $i<@suffix; ++$i) {
    # first the ps file
    $link = ($wwwdir . $expname . "_" . $sources[$nmaps] . "_" . $name .
        '.' . $suffix[$i]);
    #$link=zipit($link);
    ($start_href, $end_href) = mkhref($link);
    print OUTFILE " ";
    print OUTFILE $start_href;
    print OUTFILE "$text[$i]";
    print OUTFILE $end_href;
    if ($i+1 < scalar(@suffix)) {
      print OUTFILE ", or \n";
    }else {
      print OUTFILE ". <br />\n";
    }
  }

}

File: src/evn_postprocess/feedback.py
"""Generation of the EVN Pipeline feedback HTML page.

This is a Python port of the historical ``feedback.pl`` Perl script (originally by
Cormac Reynolds, 2001). It produces the ``{expname}.html`` feedback page that links to
the EVN pipeline output products (data summaries, plots, maps, ...) for an experiment,
marking any product missing on disk as "(not available)", and interleaves the
human-written comments stored in the ``{expname}.comment`` file.

The page skeleton lives in ``templates/feedback.html.template``; only the dynamic body
(the per-section links and comments) is assembled here.
"""
from __future__ import annotations

import os
import re
import stat
import datetime as dt
from pathlib import Path
from importlib import resources
from typing import Iterable, Sequence

from loguru import logger


# Colours reused from the original script.
_PURPLE = "#4400CC"
_RED = "#FF0000"

# Sections rendered as a single link to "{expname}_{NAME}.pdf" (optionally with a
# second descriptive sentence). Order matters: it defines the order on the page.
# Each entry is (section_name, link_text, trailing_text_or_None).
_PDF_SECTIONS: tuple[tuple[str, str, str | None], ...] = (
    ("POSSM_AUTOCORR", "Plots of the autocorrelations", None),
    ("VPLOT_UNCAL", "Plots of the uncalibrated amplitude and phase against time", None),
    ("POSSM_UNCAL", "Plots of the uncalibrated amplitude and phase against frequency channel", None),
    ("POSSM_CPOL", "The uncalibrated amplitude and phase of the crosshand correlations "
                   "against frequency channel", None),
    ("TSYS", "TSYS against time", None),
    ("GAIN", "Telescope sensitivities ", "from the a priori TSYS and Gain curves (the square of "
             "this number gives the antenna noise (SEFD) in Jy - the smaller the better)."),
    ("FRING_PHAS", "Fringe-fit phase solutions ", "(including Parallactic Angle correction)."),
    ("FRING_DELAY", "Fringe-fit delay solutions", None),
    ("FRING_RATE", "Fringe-fit rate solutions", None),
    ("BANDPASS", "Telescope bandpasses", None),
    ("VPLOT_CAL", "Calibrated amplitude and phase against time ", "(a priori amplitude "
                  "calibration and fringe-fit solutions applied)."),
    ("POSSM_CAL", "Calibrated amplitude and phase against frequency channel", None),
)


def _read_comments(comment_file: Path) -> dict[str, str]:
    """Parses a ``{expname}.comment`` file into a {section_name: comment_text} mapping.

    The file is a sequence of blocks separated by ``///`` lines. Each block starts with a
    section keyword (e.g. ``GENERAL``, ``TSYS``) followed by the free-text comment. This
    mirrors the original Perl behaviour: all lines are joined with spaces, split on
    ``///``, and each block's leading keyword is stripped to leave the comment text.
    """
    if not comment_file.exists():
        return {}

    joined = " ".join(line.rstrip("\n") for line in comment_file.read_text().splitlines())
    comments: dict[str, str] = {}
    for block in joined.split("///"):
        block = block.strip()
        if not block:
            continue
        # The keyword is the first whitespace-delimited token; the rest is the comment.
        parts = block.split(None, 1)
        name = parts[0]
        comments[name] = parts[1].strip() if len(parts) > 1 else ""
    return comments


def _guess_sources(expname: str, directory: Path) -> list[str]:
    """Returns the source names found from ``{expname}_{SOURCE}_UVPLT.pdf`` files in *directory*."""
    pattern = re.compile(rf"^{re.escape(expname)}_(.*)_UVPLT\.pdf$")
    sources: list[str] = []
    for entry in sorted(p.name for p in directory.iterdir()):
        match = pattern.match(entry)
        if match:
            sources.append(match.group(1))
    return sources


def _mkhref(link: str, directory: Path, text: str) -> str:
    """Returns an ``<a href>`` snippet for *text* pointing at *link*.

    If the file does not exist on disk (resolved relative to *directory*), the link is
    replaced by the text followed by a red "(not available)" marker, exactly as the
    original script did.
    """
    target = directory / link
    if target.exists():
        return f'<a href ="{link}">\n{text}</a> \n'
    return f'{text}<font color="{_RED}"> (not available) </font> \n'


def _comment_block(name: str, comments: dict[str, str]) -> str:
    """Renders the trailing "Comments." block for a section (closes <p>, adds <hr>)."""
    return (f'<font color="{_PURPLE}">\n'
            "Comments. <br /> \n"
            "</font>\n"
            f"{comments.get(name, '')}\n"
            "<br /> \n</p> \n<hr /> \n \n<p> \n")


def _pdf_section(expname: str, directory: Path, name: str, text: str,
                 text2: str | None, comments: dict[str, str]) -> str:
    """Renders a single-link section pointing at ``{expname}_{name}.pdf``."""
    link = f"./{expname}_{name}.pdf"
    out = _mkhref(link, directory, text)
    if text2:
        out += text2
    out += "<br />\n"
    out += _comment_block(name, comments)
    return out


def _map_links(expname: str, directory: Path, source: str, name: str,
               suffixes: Sequence[str], texts: Sequence[str]) -> str:
    """Renders the per-source list of links ``{expname}_{source}_{name}.{suffix}``."""
    out = ""
    n = len(suffixes)
    for i, (suffix, text) in enumerate(zip(suffixes, texts)):
        link = f"./{expname}_{source}_{name}.{suffix}"
        out += " " + _mkhref(link, directory, text)
        out += ", or \n" if i + 1 < n else ". <br />\n"
    return out


def _map_section(expname: str, directory: Path, sources: Sequence[str], name: str,
                 heading: str, suffixes: Sequence[str], comments: dict[str, str],
                 show_source_label: bool, texts: Sequence[str] | None = None,
                 always_comment: bool = False) -> str:
    """Renders a per-source map section.

    Args:
        heading: Introductory sentence printed before the per-source links.
        suffixes: File suffixes to link for each source (e.g. ``("pdf", "FITS")``).
        show_source_label: If True, prepend ``"{source}:"`` before each source's links
            (used when the link texts are the suffixes themselves). If False, the link
            text is the source name instead.
        texts: Explicit link texts (one per suffix). If None, the suffixes are used as the
            link texts.
        always_comment: If True, always append the comment block; otherwise only when at
            least one source was rendered (matching the original ``if nmaps > 0`` guard).
    """
    out = heading
    for source in sources:
        if show_source_label:
            out += f"{source}:"
            link_texts = list(texts) if texts is not None else list(suffixes)
        else:
            # The link text is the source name (one link per suffix).
            link_texts = [source] * len(suffixes)
        out += _map_links(expname, directory, source, name, suffixes, link_texts)

    if sources or always_comment:
        out += _comment_block(name, comments)
    return out


def _build_body(expname: str, directory: Path, sources: Sequence[str],
                comments: dict[str, str]) -> str:
    """Assembles the dynamic HTML body (everything between the intro and the footer)."""
    parts: list[str] = []

    # --- GENERAL section: data summary + scan listing links, then general comments. ---
    general = (f'<p>\n<font color="{_PURPLE}">\nGeneral Comments. \n</font>\n(')
    general += _mkhref(f"./{expname}.DTSUM", directory, "Brief data summary")
    general += " and "
    general += _mkhref(f"./{expname}.SCAN", directory, "scan listing")
    general += ") <br />\n"
    general += comments.get("GENERAL", "")
    general += "<br /> \n</p>  \n"
    parts.append(general)

    parts.append("\n<hr /> \n \n<p> \n")

    # --- Simple PDF sections. ---
    for name, text, text2 in _PDF_SECTIONS:
        parts.append(_pdf_section(expname, directory, name, text, text2, comments))

    # --- Per-source map sections. ---
    parts.append(_map_section(
        expname, directory, sources, "IMAPN",
        "Naturally weighted dirty map (not useful for bright sources) produced before "
        "self-cal of: <br />", ("pdf", "FITS"), comments, show_source_label=True))

    parts.append(_map_section(
        expname, directory, sources, "IMAPU",
        "Uniformly weighted dirty map (not useful for bright sources) produced before "
        "self-cal of: <br />", ("pdf", "FITS"), comments, show_source_label=True))

    parts.append(_map_section(
        expname, directory, sources, "CALIB_PHAS1",
        "Phase corrections applied to a priori calibrated and fringe-fitted data by "
        "self-calibration. <br />\n", ("pdf",), comments, show_source_label=False,
        always_comment=True))

    parts.append(_map_section(
        expname, directory, sources, "CALIB_AMP2",
        "Amplitude corrections applied to a priori calibrated and fringe-fitted data by "
        "self-calibration. <br />\n", ("pdf", "TXT", "ampcal"), comments,
        show_source_label=True, texts=("pdf", "text file", "statistical summary"),
        always_comment=True))

    parts.append(_pdf_section(
        expname, directory, "SENS", "Telescope sensitivities",
        " (the total AMP gain applied during both a priori and self calibration; the "
        "square of this number gives the antenna noise (SEFD) in Jy).\n", comments))

    parts.append(_map_section(
        expname, directory, sources, "CLPHS",
        "Residual closure phase (visibility closure phase with model closure phase "
        "subtracted) for: <br />", ("pdf",), comments, show_source_label=False))

    parts.append(_map_section(
        expname, directory, sources, "VPLOT_MODEL",
        "Calibrated visibilities and the source model of: <br />", ("pdf",), comments,
        show_source_label=False))

    parts.append(_map_section(
        expname, directory, sources, "UVPLT",
        "Calibrated visibilities against <em>u,v</em> distance for: <br />",
        ("pdf", "png"), comments, show_source_label=True))

    parts.append(_map_section(
        expname, directory, sources, "UVCOV",
        "<em>u,v</em> coverage for: <br />", ("pdf", "png"), comments,
        show_source_label=True))

    parts.append(_map_section(
        expname, directory, sources, "ICLN",
        "<em>Crude</em> maps of sources: <br />", ("pdf", "FITS"), comments,
        show_source_label=True))

    parts.append("</p>\n\n")
    return "".join(parts)


def generate_feedback_page(expname: str, sources: Iterable[str] | None = None,
                           nme: bool = False, contact: str | None = None,
                           directory: str | Path = ".") -> Path:
    """Generates the ``{expname}.html`` pipeline-feedback page.

    Port of ``feedback.pl``. Links every pipeline output product for the experiment,
    marks the missing ones as "(not available)", and embeds the comments from
    ``{expname}.comment``.

    Args:
        expname: Experiment name / file prefix (e.g. ``"n26k2"`` or ``"n26k2_1"`` for a
            given pipeline pass). Used verbatim to build the output product file names.
        sources: Source names to render in the per-source sections. If None, they are
            guessed from the ``{expname}_{SOURCE}_UVPLT.pdf`` files present in *directory*.
        nme: If True, format the page for a Network Monitoring Experiment; otherwise for a
            user experiment (only affects the title and heading).
        contact: Support scientist contact e-mail. A bare username (no ``@``) is completed
            with ``@jive.eu``. If None, no specific address requirement is enforced.
        directory: Directory containing the pipeline products and where the ``.html`` file
            is written. Defaults to the current directory.

    Returns:
        Path: The path to the written HTML file.
    """
    directory = Path(directory)

    if contact and "@" not in contact:
        contact = f"{contact}@jive.eu"

    # Resolve and normalise the source list (always upper-cased, as in the original).
    if sources is None:
        source_list = _guess_sources(expname, directory)
    else:
        source_list = list(sources)
    source_list = [s.upper() for s in source_list]

    comments = _read_comments(directory / f"{expname}.comment")

    title = ("European VLBI Network NME Feedback" if nme
             else "European VLBI Network Pipeline Feedback")
    heading = ("EVN Network Monitoring Experiment (NME) Pipeline Feedback" if nme
               else "EVN User Experiment Pipeline Feedback")
    # ``date``-like timestamp, kept on a single line as the original `date` output.
    last_updated = dt.datetime.now().strftime("%a %b %d %H:%M:%S %Y")

    body = _build_body(expname, directory, source_list, comments)

    template = resources.files("evn_postprocess.templates").joinpath(
        "feedback.html.template").read_text()
    replacements = {
        "{title}": title,
        "{heading}": heading,
        "{expname_upper}": expname.upper(),
        "{last_updated}": last_updated,
        "{contact}": contact or "",
        "{body}": body,
    }
    for placeholder, value in replacements.items():
        template = template.replace(placeholder, value)

    htmlfile = directory / f"{expname}.html"
    htmlfile.write_text(template)
    # Match the original `chmod 0755`.
    os.chmod(htmlfile, stat.S_IRWXU | stat.S_IRGRP | stat.S_IXGRP | stat.S_IROTH | stat.S_IXOTH)

    logger.info(f"Created pipeline feedback page: {htmlfile}")
    return htmlfile
